src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+    wconjh = FEFunction(fem_params.U, conj(w_vec)) 
+    
+    l_temp(dp) = ∫(real(-2 * DAdpf(uh, wconjh, pfh; phys_params, β, η)) * dp)fem_params.dΩ_d
+    dgfdpf = assemble_vector(l_temp, fem_params.Pf)
+    return dgfdpf
+end
+
+# Next we define the relation between $p_f$ and $p$, and get the derivative of the filter also using adjoint method.
+# 
+
+function pf_p0(p0; r, fem_params)
+    pf_vec = Filter(p0; r, fem_params)
+    pf_vec[pf_vec .< 0] .= 0
+    pf_vec[pf_vec .> 1] .= 1.0
+    pf_vec
+end
+
+function rrule(::typeof(pf_p0), p0; r, fem_params)
+  function pf_pullback(dgdpf)
+    NO_FIELDS, Dgdp(dgdpf; r, fem_params)
+  end
+  pf_p0(p0; r, fem_params), pf_pullback
+end
+
+function Dgdp(dgdpf; r, fem_params)
+    Af = assemble_matrix(fem_params.Pf, fem_params.Qf) do u, v
+        ∫(a_f(r, u, v))fem_params.dΩ_d + ∫(v * u)fem_params.dΩ_d
+    end
+    wvec = Af' \ dgdpf
+    wh = FEFunction(fem_params.Pf, wvec)
+    l_temp(dp) = ∫(wh * dp)fem_params.dΩ_d
+    return assemble_vector(l_temp, fem_params.P)
+end
+
+# Finally, we pack up everything and rewrite it as a compact form. We also record every optimization step values.
+# 
+
+function gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    pf_vec = pf_p0(p0; r, fem_params)
+    gf_pf(pf_vec; β, η, phys_params, fem_params)
+end
+
+function gf_p(p0::Vector, grad::Vector; r, β, η, phys_params, fem_params)
+    if length(grad) > 0
+        dgdp, = Zygote.gradient(p -> gf_p(p; r, β, η, phys_params, fem_params), p0)
+        grad[:] = dgdp
+    end
+    gvalue = gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    open("gvalue.txt", "a") do io
+        write(io, "$gvalue \n")
+    end
+    gvalue
+end
+
+# Using the following codes, we can check if we can get the derivatives correctly from the adjoint method by comparing it with the finite difference results.
+# 
+
+p0 = rand(fem_params.np)
+δp = rand(fem_params.np)*1e-8
+grad = zeros(fem_params.np)
+
+g0 = gf_p(p0, grad; r, β, η, phys_params, fem_params)
+g1 = gf_p(p0+δp, []; r, β, η, phys_params, fem_params)
+g1-g0, grad'*δp
+
+# ## Optimization with  NLop
+# 
+# Now we use NLopt.jl package to implement the MMA algorithm for optimization. Note that we start with $\beta=8$ and then gradually increase it to $\beta=32$ in consistant with Ref. [6].
+# 
+
+using NLopt, DelimitedFiles
+
+function gf_p_optimize(p_init, r, β, η, TOL = 1e-4, MAX_ITER = 500; phys_params, fem_params)
+    ##################### Optimize #################
+    opt = Opt(:LD_MMA, fem_params.np)
+    lb = zeros(fem_params.np)
+    ub = ones(fem_params.np)
+    opt.lower_bounds = lb
+    opt.upper_bounds = ub
+    opt.ftol_rel = TOL
+    opt.maxeval = MAX_ITER
+    opt.max_objective = (p0, grad) -> gf_p(p0, grad; r, β, η, phys_params, fem_params)
+    if (length(p_init)==0)
+        p_initial_guess = readdlm("p_opt_value.txt", Float64)
+        p_initial_guess = p_initial_guess[:]
+    else
+        p_initial_guess = p_init[:]
+    end
```

Comment:
I wouldn't do this here.   If the user wants to read the initial parameters from a file, they should call `readdlm` themselves, not pass a length-0 array.

Reply:
The thing is that we change β during optimization. For β = 8 we start with an initial guess. But for β = 16, 32, we need to read parameters from the previous optimized β. Of course we can add a new parameter to tell the function to read from the file or not, but I think we can make things simpler this way....